### DISCUSSION TITLE
Using Grott Proxy and Grott Server together

### DISCUSSION THREAD

**Discussion by matthewmcneill** [johanmeijer/grott#598] [Ideas]
I was thinking about a setup that would be more effective in isolating commands from the Growatt servers.  What I was thinking is that it would be good to chain the proxy and server together.  Basically I want to continue to share data to the Growatt servers like the grottproxy does today.  However I would like to command the Inverter using grottserver.   

Ideally we could take all the commands from the Growatt servers and queue them somehow for manual release through the server, where we could set more fine-grained rules for 
- automatic pass-through of commands, 
- manual unpacking, review and passthrough of commands following user review
- blocking and preventing commands
- identifying unknown command structures
- redirecting command such as reconfiguring the proxy connection (e.g. server IP address should reconfigure the proxy, but not the shinewifi stick)

Has there been any thinking around this sort of use-case?

M

**Comment by johanmeijer** [comment]
In the next release (3.0)  Grott and Grottserver will be integrated.  In principle Grottserver will do al communication but there will be a capability to send the data (passthrough) to growatt. That will be one way traffic (from grott/grottserver to growatt) everything (commands) from growatt will be blocked,  

This release is almost finished (alpha version) and is in test phase.

**Reply by matthewmcneill** [reply]
Oh, super interesting!  I'll look forward to testing that.

M

**Comment by johanmeijer** [comment]
It is comming.  Please give me some time. 
I might have tried to do to much changes at once and need some additional test and debug time :(

**Reply by matthewmcneill** [reply]
Keep up the good work, it's really amazing what you have achieved so far.  I'm just getting to understand the code a bit as I'm playing with grott running on my home assistant setup.  

I will take a look when you publish and perhaps do some documentation of the code as I learn more about the code base.

M

**Comment by jmalm** [comment]
I am also looking forward to this functionality. I have actually started on an implementation of my own, but will probably wait for yours. 

Would you like some help with testing? I have a MOD-10KTL3-XH, with ARK batteries that I'm aiming to control by setting Time-of-Use from home assistant.

**Comment by PedroKTFC** [comment]
I run Grott and Grottserver separately (admittedly my customised version) and have done so most of the year. I haven't communicated with the Growatt servers for many months. It might be worth you doing that to get it working before the fully integrated version. It's a great learning experience!

**Reply by jmalm** [reply]
I have used Growatt server a bit, and found that I can control the time of use that way. I have thought about skipping the Growatt server altogether, but I kind of like having that interface available. It also makes it easier to have the manufacturer "truth" easily accessible when developing software to control the inverter. 

@johanmeijer, is the 3.0-dev branch available anywhere?

**Comment by johanmeijer** [comment]
I will publish the 3.0 version in alpha (3.0.0) this week. Grott 3.0 is Grott and Grottserver integrated. 

One optie of Grott 3.0 is the server passthrough mode. Grott will act as grottserver but the data is also send to Growatt. The data from Growatt will be blocked totally.

**Reply by jmalm** [reply]
Great! Looking forward to trying it out.

**Comment by PedroKTFC** [comment]
For many months I've been running my system using Grott and grottserver. I've recently noticed that the time in my inverter (not my data logger - the two seem to be distinct) has drifted by 2-3 minutes. Is there a way of resetting the inverter's time?

**Reply by jmalm** [reply]
There is probably a register for that. I can try to investigate that for the MOD 10KTL3-XH. Maybe it will be the same register for your inverter. Will have to modify the code a bit to decode the message from the server when sending the time from the ShinePhone app.

**Reply by PedroKTFC** [reply]
Indeed. I don't know how to find out which register it would be so hope yours is the same as mine!

**Reply by jmalm** [reply]
Alright. On my inverter, the time is in registers 45 - 50, as follows:
45: year EDIT: when _setting_ the time, this should be number of years since 2000, i.e. 24 for year 2024. When reading, it's from year 0.
46: month
47: day
48: hour
49: minute
50: second

Data carrying part of a message response after setting or asking to read those registers:

`002d003207e8000b001a0016000f000c`

The first four bytes is the from-register and to-register. The register values immediately follow. The above is translated to 2024-11-26 22:15:12.

**Reply by PedroKTFC** [reply]
Wow! How did you work that out? I've just looked at my inverter and these registers align with mine. I can see my inverter is about 4 minutes out. However, although I can set other (4 digit) register values, I can't set the minute value (which would be sufficient for me). I'm guessing that you have to set all 6 registers in one go. I'll have to work out how to do that! 😭

**Reply by jmalm** [reply]
I modified grottproxy.py at the end, to print out all messages (if a special file exists, but that's just a practical way to start / stop the printout):

```python
        import os
        if os.path.exists("print_all_messages"):
            print(f"Message received: {data}")
```

I then set the time from the ShinePhone app, and watched the messages output from grott (my grott is set up to forward to and recieve from the growatt server). Having previously structured @johanmeijer's reverse-engineered message parsing in grottdata.py into separate classes in my fork (https://github.com/jmalm/grott/tree/proxy-api/messages), I was able to use those to understand the data in the messages.

Oh, and I use your implementation of libscrc in that branch too. Great work! Open source is a beautiful thing.

**Reply by PedroKTFC** [reply]
Well that was inspired, well done! I'm in the process of updating my version of grottserver to send an update time message. I'm assuming it'll be a standard update multiple registers message. I'll be simply doing an update to `now()` using the inverter datetime option.

**Reply by PedroKTFC** [reply]
I've changed my grottserver to send what I think should be the update message but it's returning an error from the inverter. The body of the message looks the same as yours other than the time is for "now". Are you able to extract the full message the growatt server sends to update the time so I can check where mine is going wrong? Thanks.

**Reply by jmalm** [reply]
I would, but the raw message includes my inverter ID, which I don't think is wise to spread on the internet. Let me see if I can construct a message that works, and then replace the inverter ID. Or, even better perhaps, send the code to generate the message. I may be able to take a shot at it tonight. It may already be in grottserver.py, but I think I remember having seen that there were still some issues with sending multiregister put requests. Were you involved in that?

**Reply by PedroKTFC** [reply]
I've emailed you directly. Hope you got it.

**Reply by jmalm** [reply]
Alright, now there is an implementation of `messages.inverter_put.InverterPutMultipleCommand` that should serialize into a working message to send to the server, on https://github.com/jmalm/grott/commits/proxy-api/.

The following should return `True`, if `raw_from_growatt` is a raw message to set inverter time, sent from the growatt server:

```python
import messages
messages.deserialize(raw_from_growatt).serialize() == raw_from_growatt
```

To create a message from scratch:

```python
import messages
messages.InverterPutMultipleCommand(protocol=6, device_id1, datalogger_id=b'abcd123456', start_register=45, end_register=50, values=[24, 11, 27, 12, 39, 10]).serialize()

Out: b'\x00\x01\x00\x06\x000\x01\x10&\x10\x0c\x13PFGsGYwattGrowattGrowattGroZaFt_rdwztxGUo}y\xc3'
```

NOTE: the year in the command from the growatt server appears to be number of years from 2000. I don't know if putting 2024 instead of 24 will also work.

I haven't tested sending this with grottserver, as I don't have it running currently.

(I realize now, that I shouldn't need to provide `end_register` to the constructor, since the number of items in `values` should give that.)

**Reply by PedroKTFC** [reply]
Great stuff. Using the number of years from 2000 seems to have cured my error! I've now updated my version of grottserver so that when reading all registers it also reads the inverter time registers and, if you use the datetime option with the inverter PUT, it will update the inverter time to "now" when executing the PUT.

**Reply by jmalm** [reply]
It's funny, because the message that is sent by the inverter, as a response to the message that sets the time, has the full year, 2024! 😵‍💫

**Reply by PedroKTFC** [reply]
Indeed. That's what stopped me setting the register when I sent the year 2024. Had it worked and the year been 4024 I'd have spotted the issue! 😆

**Reply by johanmeijer** [reply]
I missed this one It think. Start with a reply on the original question. 

Stop/starting grott will update the Inverter time (with the time from growatt if you run Grott standalone, with the time from your server if you run grottserver). 

The grottserver API also allow to set invert time to server time with a command: 

http put : \<yougrottserverrip\>:5782/datalogger?command=datetime&datalogger=\<yourdatalogger\>

**Reply by PedroKTFC** [reply]
I missed this one too! Setting the data logger time doesn't seem to set the inverter time. In fact my data logger time is often out (I'm not sure what's going on with it tbh). However, the inverter time seems to be much more robust (it does slowly drift over a period of weeks/months). That's why I implemented a set inverter option in my grottserver fork. This fork includes what I think are other useful features such as setting time slots (see https://github.com/johanmeijer/grott/issues/538).

**Reply by johanmeijer** [reply]
Maybe I did tell things wrong/unclear. Grottserver does update the datalogger time only (and only when the inverter is started/reconnect). This is how it also works when connected to the growatt servers. 

I am not sure if and how the Inverter time is set  automatically by a "normal" connection with the growatt servers. For my old tl-s inverters the inverter time is not set automatically. For me it is not very important while nothing is done with this time at this moment. I can understand when you have a more modern inverter with batteries connected or when you have to pay for the electricity you deliver to grid you want to use this. 

You have to set inverter registers to change the inverter time (I think register 45-50, 51 seems for the newer inverter also being used as  week number, not in there for the TL-S). 

We can make a command for this in grottserver (grott 3.0). 

Question for the people with the new inverters: Is your time automatically being set by growatt? If yes is it possible to find the command for this in the log?

**Reply by PedroKTFC** [reply]
No worries. You've done a marvellous job, especially when it's not your first language.

Btw, I have created a pull request https://github.com/johanmeijer/grott/pull/612 to document registers 45-50.

**Comment by johanmeijer** [comment]
Grott 3.0.0 now available in the alfa branche. Please let me know your opinion (discussions https://github.com/johanmeijer/grott/discussions/613)

**Reply by jmalm** [reply]
Awesome! Will check it out and get back to you.

**Reply by PedroKTFC** [reply]
I've had a quick look at the grottserver code. Unfortunately I can't use that version. My version has diverged quite significantly from yours. In particular, there are a couple of functions I now use that are essential to me, namely: reading the inverter registers (I do this frequently to check what they're set to and to confirm they're set as I want) and setting the triple set of time slot registers (these can only be set as a triple). On the other hand, Grott itself should not present a problem.

**Reply by johanmeijer** [reply]
@PedroKTFC yes a lot of code has been changed and will change with the next builds (that is why I started a new version).  
The good news for now is that I did not touched the HTTPSERVER part yet. So if you have made changes in that area it would easy to fit that to the 3.0.0. build.

**Comment by PedroKTFC** [comment]
I'm away for a week so I'll look at it when I'm back. Btw, I run it as an
HA add-on, is this version available as one or will I have to run it
locally to test it?

**Comment by johanmeijer** [comment]
I do think there is already a HA version yet for 3.0.0 (it is brand new and @egguy does not had time to make this). So you have to test it locally at this moment.

**Reply by egguy** [reply]
Hi, I will take a look. Do you think the reuse of the `-beta` branch could be appropriate?

**Reply by johanmeijer** [reply]
Hi @egguy, if we think the alpha is stable enough we can promote it to beta.  
Lets wait a week and then do this if there are no complains ;)

**Comment by jmalm** [comment]
I've used it a bit, and it seems to be working fine. Thanks for all the hard work that went into this @johanmeijer !

One thing to note: You have to set the growatt server using the IP address (not `server.growatt.com`), or the blocking doesn't work (it compares IP address and port).

I've added a small functionality to the http server to collect messages sent from the growatt server. It's available on https://github.com/jmalm/grott/tree/collect-messages and works like this:

Start collecting messages from the growatt server (after 10 messages have been collected, no more messages are collected):
```bash
curl -X PUT http://localhost:5782/collect_messages?max=10
```

Get the collected messages:
```bash
curl -X GET http://localhost:5782/collect_messages
# Returns a JSON list with the collected messages, both as "raw" data, and parsed, if the message is known. The first of the following is the message to read the values from inverter registers 45 - 50 (inverter date and time).
[
  {
    "message": {
      "type": "InverterGetMulti",
      "send_sequence": 1,
      "protocol": 6,
      "device_id": 1,
      "record_type": 5,
      "_body_length": 36,
      "datalogger_id": "my logger id",
      "start_register": 45,
      "end_register": 50
    },
    "datetime": "2024-12-11T11:37:01.491935",
    "address": "47.254.130.145",
    "port": 5279,
    "from_datalogger": false,
    "raw": "0001000600240105.......1747447726f5a614630fb"
  },
  {
    "message": {
      "type": "Message",
      "send_sequence": 42,
      "protocol": 6,
      "device_id": 1,
      "record_type": 3,
      "_body_length": 3
    },
    "datetime": "2024-12-11T11:37:14.723489",
    "address": "47.254.130.145",
    "port": 5279,
    "from_datalogger": false,
    "raw": "002a00060003010347b799"
  }
]
```

Stop collecting messages:
```
curl -X DELETE http://localhost:5782/collect_messages
```

The parsing of messages is done by a `messages` module that I think could become something really good to use in the rest of grott. It serializes and deserializes growatt messages. I have been able to figure out quite a few messages, but there are many left to implement.

I plan to use this to map the messages and registers for the MOD 10KTL3-XH inverter, so that I can manage the battery through some automation. For this purpose, the messages that I've implemented already are probably enough. It's basically just the register mapping that's left.

@johanmeijer , if you would like to include it upstream, I would be happy to prepare a pull request. An alternative approach could be to make a separate growatt messages repo (not sure what to call it - I don't think I can/should use "growatt" in the name), that could be used in the grott repo. I think it could simplify the grottserver code a bit.

**Reply by jmalm** [reply]
Oh, and by the way, I didn't see a need to update anything in Home Assistant. Seems to be working correctly with the Alpha branch. Is there anything specific that I should be looking out for?

**Reply by johanmeijer** [reply]
I am going to look at  it. The HTTP part is the part I have not changed yet.
